File: 2016/21_ScrambledLettersandHash/pword.ts
// ======================================================================
// Scrambled Letters and Hash
//   Advent of Code 2016 Day 21 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           p w o r d . t s
//
// A solver for the Advent of Code 2016 Day 21 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                 import
// ----------------------------------------------------------------------

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type Operand = 'a' | 'b' | 'c' | 'd' | 'e' | 'f' | 'g' | 'h' | '0' | '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | 'X';
type Actions = 'swap position'
| 'swap letter'
| 'rotate left'
| 'rotate right'
| 'rotate based'
| 'reverse positions'
| 'move position';
type Instruction = {
  action: Actions,
  op1: Operand,
  op2: Operand,
};
type Instructions = Instruction[];
type Op2num = Record<Operand, number>;

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const UNUSED = 'X';
const START = 'abcdefgh';
const FINISH = 'fbgdceah';

const OP2NUM: Op2num = {
  a: -1,
  b: -1,
  c: -1,
  d: -1,
  e: -1,
  f: -1,
  g: -1,
  h: -1,
  0: 0,
  1: 1,
  2: 2,
  3: 3,
  4: 4,
  5: 5,
  6: 6,
  7: 7,
  8: 8,
  9: 9,
  X: -1,
};

const UNSCRAMBLE_ROTATE_BASED_5 = [1, 3, 0, 2, 4];
const UNSCRAMBLE_ROTATE_BASED_8 = [1, 1, 6, 2, 7, 3, 0, 4];
// ======================================================================
//                                                                  Pword
// ======================================================================

export class Pword {
  // Object for Scrambled Letters and Hash
  text: string[];

  part2: boolean;

  instructions: Instructions;

  start: string;

  finish: string;

  constructor(text: string[], part2 = false) {
    // Create a Pword object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.instructions = [];
    this.start = START;
    this.finish = FINISH;

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText();
    }
  }

  processText(): void {
    // 1. Start with nothing
    this.instructions = [];
    // 2. Loop for all lines of the text
    this.text.forEach((line) => {
      // 3. Split the line up
      const parts = line.split(' ');
      // 4. Decode based on the first word
      switch (parts[0]) {
        case 'swap':
          // 4a. swap letter l1 l2 and position p1 p2
          switch (parts[1]) {
            case 'letter':
              this.instructions.push({
                action: 'swap letter',
                op1: <Operand>parts[2],
                op2: <Operand>parts[5],
              });
              break;
            case 'position':
              this.instructions.push({
                action: 'swap position',
                op1: <Operand>parts[2],
                op2: <Operand>parts[5],
              });
              break;
            default:
              console.log(`Unexpected swap instruction ${line}`);
              break;
          }
          break;
        case 'rotate':
          // 4b. rotate left n1, right n1, and based l1
          switch (parts[1]) {
            case 'left':
              this.instructions.push({
                action: 'rotate left',
                op1: <Operand>parts[2],
                op2: UNUSED,
              });
              break;
            case 'right':
              this.instructions.push({
                action: 'rotate right',
                op1: <Operand>parts[2],
                op2: UNUSED,
              });
              break;
            case 'based':
              this.instructions.push({
                action: 'rotate based',
                op1: <Operand>parts[6],
                op2: UNUSED,
              });
              break;
            default:
              console.log(`Unexpected rotate instruction ${line}`);
              break;
          }
          break;
        case 'reverse':
          // 4c. reverse positions p1 p2
          this.instructions.push({
            action: 'reverse positions',
            op1: <Operand>parts[2],
            op2: <Operand>parts[4],
          });
          break;
        case 'move':
          // 4d move positions p1 p2
          this.instructions.push({
            action: 'move position',
            op1: <Operand>parts[2],
            op2: <Operand>parts[5],
          });
          break;
        default:
          console.log(`Unexpected instruction ${line}`);
      }
    });
  }

  static execute(inst: Instruction, input: string): string {
    const result = input.split('');
    let swap = '?';
    let index1 = OP2NUM[inst.op1];
    let index2 = OP2NUM[inst.op2];
    // console.log(`execute ${index1} ${index2} ${inst.action} ${inst.op1} ${inst.op2}`);
    let subArray: string[] = [];
    switch (inst.action) {
      case 'swap position':
        swap = result[index1];
        result[index1] = result[index2];
        result[index2] = swap;
        break;
      case 'swap letter':
        index1 = result.indexOf(inst.op1);
        index2 = result.indexOf(inst.op2);
        swap = result[index1];
        result[index1] = result[index2];
        result[index2] = swap;
        break;
      case 'rotate left':
        for (let index = index1; index > 0; index -= 1) {
          result.push(<string>result.shift());
        }
        break;
      case 'rotate right':
        for (let index = index1; index > 0; index -= 1) {
          result.unshift(<string>result.pop());
        }
        break;
      case 'rotate based':
        // - rotate based on position of letter X means that the whole string
        //   should be rotated to the right based on the index of letter X
        //   (counting from 0) as determined before this instruction does any
        //   rotations. Once the index is determined, rotate the string to the
        //   right one time, plus a number of times equal to that index, plus
        //   one additional time if the index was at least 4.
        // - rotate based on position of letter b finds the index of letter b (1),
        //   then rotates the string right once plus a number of times equal to
        //   that index (2)
        //      abdec --> ecabd.
        //  - rotate based on position of letter d finds the index of letter d (4),
        //    then rotates the string right once, plus a number of times equal to
        //    that index, plus an additional time because the index was at least 4,
        //    for a total of 6 right rotations
        //      ecabd --> decab.
        index1 = result.indexOf(inst.op1);
        index1 += index1 >= 4 ? 2 : 1;
        for (let index = index1; index > 0; index -= 1) {
          result.unshift(<string>result.pop());
        }
        break;
      case 'reverse positions':
        subArray = result.slice(index1, index2 + 1);
        for (let index = index1; index <= index2; index += 1) {
          result[index] = <string>subArray.pop();
        }
        break;
      case 'move position':
        // - move position X to position Y means that the letter which is at index
        //   X should be removed from the string, then inserted such that it ends
        //   up at index Y.
        // - move position 1 to position 4 removes the letter at position 1 (c),
        //   then inserts it at position 4 (the end of the string)
        //     bcdea --> bdeac.
        // - move position 3 to position 0 removes the letter at position 3 (a),
        //   then inserts it at position 0 (the front of the string)
        //     bdeac --> abdec.
        swap = result[index1];
        if (index1 < index2) {
          for (let index = index1; index < index2; index += 1) {
            result[index] = result[index + 1];
          }
        } else if (index1 > index2) {
          for (let index = index1; index > index2; index -= 1) {
            result[index] = result[index - 1];
          }
        }
        result[index2] = swap;
        break;
      default:
        console.log(`Unable to execute instruction ${inst.action} ${inst.op1} ${inst.op2}`);
        break;
    }
    return result.join('');
  }

  scramble(verbose = false, limit = 0): string {
    if (verbose) console.log(`scramble of ${this.start}, limit=${limit}`);
    // 1. Start at the start
    let result = this.start;
    // 2. Loop for all of the instructions
    this.instructions.forEach((inst) => {
      // 3. Execute the instruction
      const next = Pword.execute(inst, result);
      if (verbose) console.log(`scramble ${result} --> ${next} inst ${inst.action} ${inst.op1} ${inst.op2}`);
      // 4. The scrample is now the result
      result = next;
    });
    // 5. Return the final scrambled result
    return result;
  }

  static reverse(inst: Instruction, input: string): string {
    const result = input.split('');
    let swap = '?';
    let index1 = OP2NUM[inst.op1];
    let index2 = OP2NUM[inst.op2];
    // console.log(`reverse ${index1} ${index2} ${inst.action} ${inst.op1} ${inst.op2}`);
    let subArray: string[] = [];
    switch (inst.action) {
      case 'swap position': // reverse is same as execute
        swap = result[index1];
        result[index1] = result[index2];
        result[index2] = swap;
        break;
      case 'swap letter': // reverse is same as execute
        index1 = result.indexOf(inst.op1);
        index2 = result.indexOf(inst.op2);
        swap = result[index1];
        result[index1] = result[index2];
        result[index2] = swap;
        break;
      case 'rotate left': // reverse is rotate right
        for (let index = index1; index > 0; index -= 1) {
          result.unshift(<string>result.pop());
        }
        break;
      case 'rotate right': // reverse if rotate left
        for (let index = index1; index > 0; index -= 1) {
          result.push(<string>result.shift());
        }
        break;
      case 'rotate based':
        // reverse has left rotate instead of right
        // but count determined empirically
        index1 = result.indexOf(inst.op1);
        index1 = result.length === 8 ? UNSCRAMBLE_ROTATE_BASED_8[index1]
          : UNSCRAMBLE_ROTATE_BASED_5[index1];
        for (let index = index1; index > 0; index -= 1) {
          result.push(<string>result.shift());
        }
        break;
      case 'reverse positions': // reverse is same as execute
        subArray = result.slice(index1, index2 + 1);
        for (let index = index1; index <= index2; index += 1) {
          result[index] = <string>subArray.pop();
        }
        break;
      case 'move position': // reverse swaps the indexes
        // - move position X to position Y means that the letter which is at index
        //   X should be removed from the string, then inserted such that it ends
        //   up at index Y.
        // - move position 1 to position 4 removes the letter at position 1 (c),
        //   then inserts it at position 4 (the end of the string)
        //     bcdea --> bdeac.
        // - move position 3 to position 0 removes the letter at position 3 (a),
        //   then inserts it at position 0 (the front of the string)
        //     bdeac --> abdec.
        index2 = OP2NUM[inst.op1];
        index1 = OP2NUM[inst.op2];
        swap = result[index1];
        if (index1 < index2) {
          for (let index = index1; index < index2; index += 1) {
            result[index] = result[index + 1];
          }
        } else if (index1 > index2) {
          for (let index = index1; index > index2; index -= 1) {
            result[index] = result[index - 1];
          }
        }
        result[index2] = swap;
        break;
      default:
        console.log(`Unable to execute instruction ${inst.action} ${inst.op1} ${inst.op2}`);
        break;
    }
    return result.join('');
  }

  unscramble(verbose = false, limit = 0): string {
    if (verbose) console.log(`unscramble of ${this.finish}, limit=${limit}`);
    // 1. Start at the end
    let result = this.finish;
    // 2. Loop for all of the instructions (in reverse)
    for (let index = this.instructions.length - 1; index >= 0; index -= 1) {
      const inst = this.instructions[index];
      // 3. Execute the instruction
      const next = Pword.reverse(inst, result);
      if (verbose) console.log(`unscramble ${result} --> ${next} inst ${inst.action} ${inst.op1} ${inst.op2}`);
      // 4. The scrample is now the result
      result = next;
    }
    // 5. Return the final scrambled result
    return result;
  }

  solution(verbose = false, limit = 0): string {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      return this.unscramble(verbose, limit);
    }
    return this.scramble(verbose, limit);
  }

  partOne(verbose = false, limit = 0): string {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): string {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                         p w o r d . t s                        end
// ======================================================================

File: 2016/21_ScrambledLettersandHash/pword.py
# ======================================================================
# Scrambled Letters and Hash
#   Advent of Code 2016 Day 21 -- Eric Wastl -- https://adventofcode.com
#
# Python implementation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                         p w o r d . p y
# ======================================================================
"A solver for the Advent of Code 2016 Day 21 puzzle"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
PART_ONE_CLEAR = "abcdefgh"
PART_TWO_SCRAMBLED = "fbgdceah"
UNSCRAMBLE_ROTATE_BASED_5 = [1, 3, 0, 2, 4]
UNSCRAMBLE_ROTATE_BASED_8 = [1, 1, 6, 2, 7, 3, 0, 4]

# ======================================================================
#                                                                  Pword
# ======================================================================


class Pword(object):   # pylint: disable=R0902, R0205
    "Object for Scrambled Letters and Hash"

    def __init__(self, text=None, part2=False):

        # 1. Set the initial values
        self.part2 = part2
        self.text = text

    def scrambler(self, clear_text):
        "Scramble clear text into encrypted text"

        # 1. Start with the clear text
        result = clear_text

        # 2. Loop for each of the scrampling instructions
        for inst in self.text:

            # 3. Execute the instruction
            result = Pword.execute(inst, result)

        # 4. Returned the scrambled result
        return result

    @staticmethod
    def execute(inst, in_word):
        "Execute a single scrambler instruction"

        # 1. Split the word into letters
        letters = list(in_word)

        # 2. Break up the instruction into words
        words = inst.split(' ')

        # 3. Execute the instruction
        if words[0] == "swap" and words[1] == "position":
            # swap position X with position Y
            # means that the letters at indexes X and Y (counting from 0) should be swapped.
            x_indx = int(words[2])
            y_indx = int(words[5])
            temp = letters[x_indx]
            letters[x_indx] = letters[y_indx]
            letters[y_indx] = temp
        elif words[0] == "swap" and words[1] == "letter":
            # swap letter X with letter Y
            # means that the letters X and Y should be swapped
            #   (regardless of where they appear in the string).
            x_indx = letters.index(words[2])
            y_indx = letters.index(words[5])
            temp = letters[x_indx]
            letters[x_indx] = letters[y_indx]
            letters[y_indx] = temp
        elif words[0] == "rotate" and words[1] == "left":
            # rotate left/right X steps
            # means that the whole string should be rotated
            #   for example, one right rotation would turn abcd into dabc.
            length = len(letters)
            steps = int(words[2]) % length
            letters = letters + letters
            letters = letters[steps:steps + length]
        elif words[0] == "rotate" and words[1] == "right":
            # rotate left/right X steps
            # means that the whole string should be rotated
            #   for example, one right rotation would turn abcd into dabc.
            length = len(letters)
            steps = int(words[2]) % length
            letters = letters + letters
            letters = letters[length - steps:length - steps + length]
        elif words[0] == "rotate" and words[1] == "based":
            # rotate based on position of letter X
            # means that the whole string should be rotated to the right based on the
            # index of letter X (counting from 0) as determined before this instruction
            # does any rotations. Once the index is determined, rotate the string to the
            # right one time, plus a number of times equal to that index, plus
            # one additional time if the index was at least 4.
            length = len(letters)
            steps = letters.index(words[6])
            if steps >= 4:
                steps += 1
            steps += 1
            steps = steps % length
            letters = letters + letters
            letters = letters[length - steps:length - steps + length]
        elif words[0] == "reverse" and words[1] == "positions":
            # reverse positions X through Y
            # means that the span of letters at indexes X through Y
            # (including the letters at X and Y) should be reversed in order.
            x_indx = int(words[2])
            y_indx = int(words[4])
            rev = letters[x_indx:y_indx + 1]
            rev.reverse()
            letters = letters[0:x_indx] + rev + letters[y_indx + 1:]
        elif words[0] == "move" and words[1] == "position":
            # move position X to position Y
            # means that the letter which is at index X should be removed from the string,
            # then inserted such that it ends up at index Y.
            x_indx = int(words[2])
            y_indx = int(words[5])
            x_letter = letters[x_indx]
            del letters[x_indx]
            letters.insert(y_indx, x_letter)

        # 4. Return the slighly more scrambled
        return ''.join(letters)

    def unscrambler(self, encrypted):
        "Uncramble encrypted back to the clear text"

        # 1. Start with the encrypted
        result = encrypted

        # 2. Loop for each of the scrampling instructions
        for indx in range(len(self.text), 0, -1):
            inst = self.text[indx - 1]

            # 3. Un-Execute the instruction
            result = Pword.unexecute(inst, result)

        # 4. Returned the unscrambled result
        return result

    @staticmethod
    def unexecute(inst, in_word):
        "Reverse a single scrambler instruction"

        # 1. Split the word into letters
        letters = list(in_word)

        # 2. Break up the instruction into words
        words = inst.split(' ')

        # 3. Reverse the instruction
        if words[0] == "swap" and words[1] == "position":  # Reverse is the same
            # swap position X with position Y
            # means that the letters at indexes X and Y (counting from 0) should be swapped.
            x_indx = int(words[2])
            y_indx = int(words[5])
            temp = letters[x_indx]
            letters[x_indx] = letters[y_indx]
            letters[y_indx] = temp
        elif words[0] == "swap" and words[1] == "letter":  # Reverse is the same
            # swap letter X with letter Y
            # means that the letters X and Y should be swapped
            #   (regardless of where they appear in the string).
            x_indx = letters.index(words[2])
            y_indx = letters.index(words[5])
            temp = letters[x_indx]
            letters[x_indx] = letters[y_indx]
            letters[y_indx] = temp
        elif words[0] == "rotate" and words[1] == "left":  # Reverse is rotate right
            # rotate left/right X steps
            # means that the whole string should be rotated
            #   for example, one right rotation would turn abcd into dabc.
            length = len(letters)
            steps = int(words[2]) % length
            letters = letters + letters
            letters = letters[length - steps:length - steps + length]
        elif words[0] == "rotate" and words[1] == "right":  # Reverse is rotate left
            # rotate left/right X steps
            # means that the whole string should be rotated
            #   for example, one right rotation would turn abcd into dabc.
            length = len(letters)
            steps = int(words[2]) % length
            letters = letters + letters
            letters = letters[steps:steps + length]
        elif words[0] == "rotate" and words[1] == "based":  # left instead of right
            # rotate based on position of letter X
            # means that the whole string should be rotated to the right based on the
            # index of letter X (counting from 0) as determined before this instruction
            # does any rotations. Once the index is determined, rotate the string to the
            # right one time, plus a number of times equal to that index, plus
            # one additional time if the index was at least 4.
            indx = letters.index(words[6])
            length = len(letters)
            if length == 8:
                steps = UNSCRAMBLE_ROTATE_BASED_8[indx]
            else:
                steps = UNSCRAMBLE_ROTATE_BASED_5[indx]
            letters = letters + letters
            letters = letters[steps:steps + length]
        elif words[0] == "reverse" and words[1] == "positions":  # Same as execute
            # reverse positions X through Y
            # means that the span of letters at indexes X through Y
            # (including the letters at X and Y) should be reversed in order.
            x_indx = int(words[2])
            y_indx = int(words[4])
            rev = letters[x_indx:y_indx + 1]
            rev.reverse()
            letters = letters[0:x_indx] + rev + letters[y_indx + 1:]
        elif words[0] == "move" and words[1] == "position":  # reverse swaps the indexes
            # move position X to position Y
            # means that the letter which is at index X should be removed from the string,
            # then inserted such that it ends up at index Y.
            x_indx = int(words[5])
            y_indx = int(words[2])
            x_letter = letters[x_indx]
            del letters[x_indx]
            letters.insert(y_indx, x_letter)

        # 4. Return the slighly clearer work
        return ''.join(letters)

    def part_one(self, verbose=False, limit=0, word=PART_ONE_CLEAR):
        "Returns the solution for part one"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part one
        return self.scrambler(word)

    def part_two(self, verbose=False, limit=0, word=PART_TWO_SCRAMBLED):
        "Returns the solution for part two"

        # 0. Precondition axioms
        assert verbose in [True, False]
        assert limit >= 0

        # 1. Return the solution for part two
        return self.unscrambler(word)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                         p w o r d . p y                        end
# ======================================================================
